vcl_templates/_multivariate_tests.vcl.erb

```diff
@@ -4,58 +4,112 @@
+    } else {
+      declare local var.denominator_<%= test %> INTEGER;
 <% variants.each do |variant| -%>
-        declare local var.denominator_<%= test %>_<%= variant %> INTEGER;
-        declare local var.nominator_<%= test %>_<%= variant %> INTEGER;
-        set var.nominator_<%= test %>_<%= variant %> = std.atoi(table.lookup(<%= test.downcase %>_percentages, "<%= variant %>"));
-        set var.denominator_<%= test %> += var.nominator_<%= test %>_<%= variant %>;
+      declare local var.denominator_<%= test %>_<%= variant %> INTEGER;
+      declare local var.nominator_<%= test %>_<%= variant %> INTEGER;
+      set var.nominator_<%= test %>_<%= variant %> = std.atoi(table.lookup(<%= test.downcase %>_percentages, "<%= variant %>"));
+      set var.denominator_<%= test %> += var.nominator_<%= test %>_<%= variant %>;
 <% end -%>
 <% variants.each_with_index do |variant, idx| -%>
 <% if variant == variants.first -%>
-        set var.denominator_<%= test %>_<%= variant %> = var.denominator_<%= test %>;
+      set var.denominator_<%= test %>_<%= variant %> = var.denominator_<%= test %>;
 <% elsif variant != variants.last -%>
 <% previous_variant = variants[idx - 1] -%>
-        set var.denominator_<%= test %>_<%= variant %> = var.denominator_<%= test %>_<%= previous_variant %>;
-        set var.denominator_<%= test %>_<%= variant %> -= var.nominator_<%= test %>_<%= previous_variant %>;
+      set var.denominator_<%= test %>_<%= variant %> = var.denominator_<%= test %>_<%= previous_variant %>;
+      set var.denominator_<%= test %>_<%= variant %> -= var.nominator_<%= test %>_<%= previous_variant %>;
 <% end # if -%>
 <% end # variants.each_with_index -%>
 <% variants.each do |variant| -%>
 <% if variant == variants.first -%>
-        if (randombool(var.nominator_<%= test %>_<%= variant %>, var.denominator_<%= test %>_<%= variant %>)) {
-          set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
+      if (randombool(var.nominator_<%= test %>_<%= variant %>, var.denominator_<%= test %>_<%= variant %>)) {
+        set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
 <% elsif variant != variants.first && variant != variants.last -%>
-        } else if (randombool(var.nominator_<%= test %>_<%= variant %>, var.denominator_<%= test %>_<%= variant %>)) {
-          set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
+      } else if (randombool(var.nominator_<%= test %>_<%= variant %>, var.denominator_<%= test %>_<%= variant %>)) {
+        set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
 <% else -%>
-        } else {
-          set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
-        }
+      } else {
+        set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
+      }
 <% end -%>
 <% end # variants.each-%>
-      }
     }
+<% end # if CookielessAATest -%>
+  }
 <% end # test_config.each -%>
 <% end # ab_tests.each -%>
+}
+else if (req.http.Cookie !~ "cookies_preferences_set") {
```

Comment:
Is this `else if` (rather than `else`) to enable users to also reject cookieless AB tests via the cookie banner?

Reply:
The reason it's an `else if` is because we only want to place the user in a variant of this test if they haven't yet consented or declined cookies (via the main consent button in the header _or_ via the cookie preferences page). Once they've done either of these two actions, we no longer want this user to participate in the test.

If this statement were an `else` then we would continue to opt them in to the test regardless of whether they've set preferences, which isn't what we want to do. Appreciate that this logic isn't exactly straightforwards, it has also made me do a few double-takes.

Reply:
Thanks, this makes sense 👍